apps/labs/posts/scipy-array-api.md

```diff
@@ -0,0 +1,441 @@
+
+1. comply with the standard
+2. implement the `fft` extension
+
+After [the main PR](https://github.com/scipy/scipy/pull/19005) merged, the ability to use CuPy arrays and PyTorch Tensors
+with the standard extension functions was enabled when the experimental environment variable `SCIPY_ARRAY_API` is set to `1`.
+For example, the code in the following benchmark - the task of smoothing an image (taking a 2-D image,
+using `scipy.fftn`, zero-ing out the highest frequency components, and using `scipy.ifftn`) using CuPy
+arrays - now works:
+
+```python
+import scipy
+import cupy as cp
+cpface = cp.asarray(face)
+f = scipy.fft.fftn(cpface)
+fshift = scipy.fft.fftshift(f)
+original = cp.copy(fshift)
+fshift[354:414, 482:532] = 0
+f_ishift = scipy.fft.ifftshift(original - fshift)
+x = scipy.fft.ifftn(f_ishift)
+```
+
+On my machine, this demonstrated a ~20x performance improvment over using NumPy arrays in the same functions.
+You can see the full benchmark on [the PR](https://github.com/scipy/scipy/pull/19005).
+
+Completing this PR wasn't without difficulty, as lots of SciPy's brand new array API testing infrastructure
+needed improvements in the process.
+I contributed quite a few PRs to help make this testing infrastructure more robust, and squash some bugs;
+you can find all of my merged contributions relating to array types
+[here](https://github.com/scipy/scipy/pulls?q=is%3Apr+author%3Alucascolley+is%3Amerged+label%3A%22array+types%22).
+This was also my first ever big contribution to Open Source, so there was a large (but very rewarding!) learning
+process with Git, GitHub and collaboration with reviewers.
+In the next section, I'll dive into the process of converting `scipy.fft` and share some of the lessons learned
+for converting other submodules.
+
+---
+
+## How to Convert a Submodule
+
+I'll start by showing what the necessary changes look like, before explaining in more detail how some of the
+machinery works.
+
+For functions which are made up of pure Python and NumPy, the conversion process is really quite simple!
+It involves comparing the currently used functions/methods with those in
+[the standard API specification](https://data-apis.org/array-api/latest/API_specification/)
+and mimicking the current behaviour.
+The key step is the addition of `xp = array_namespace(x, y)`, where `x` and `y` (and so on for more inputs)
+are the input arrays.
+This line is used at the start of every converted function, and allows us to use the functions/methods which
+are compatible with the arrays without re-importing anything!
+The rest of the changes are mostly replacing uses of `np` with `xp`, alongside making adjustments where the
+standard API differs from the NumPy API.
+
+For example, here are the modifications made to `scipy.fft.fht`:
+
+```diff
+def fht(a, dln, mu, offset=0.0, bias=0.0):
++    xp = array_namespace(a)
++
+     # size of transform
+-    n = np.shape(a)[-1]
++    n = a.shape[-1]
```

Comment:
I would remove that from the diff as it could have been written like this before too.

Reply:
I actually quite like it being there since it's representative of the kind of small changes that are needed where the standard differs from the NumPy API - but happy to remove if others think so.

Reply:
Ah I see, things which are not in the standard and need a conversion. Then I would highlight that (comment or in the description.)